File: tiktok-research-api-python/tiktok_research_api/query.py
# Copyright 2024 TikTok Pte. Ltd.
#
# This source code is licensed under the MIT license found in
# the LICENSE file in the root directory of this source tree.

from .criteria import Criteria

__all__ = ["Query"]


class Query:
    def __init__(self, and_criteria=None, or_criteria=None, not_criteria=None):
        self.and_criteria = and_criteria or []
        self.or_criteria = or_criteria or []
        self.not_criteria = not_criteria or []

    def to_dict(self):
        return {
            "and": [criteria.to_dict() for criteria in self.and_criteria],
            "or": [criteria.to_dict() for criteria in self.or_criteria],
            "not": [criteria.to_dict() for criteria in self.not_criteria],
        }

File: tiktok-research-api-r/R/Query.R
# Copyright 2024 TikTok Pte. Ltd.
#
# This source code is licensed under the MIT license found in
# the LICENSE file in the root directory of this source tree.

#' Query Reference Class
#'
#' A class to represent a query with AND, OR, and NOT criteria.
#'
#' @field and_criteria A list of Criteria objects for AND conditions.
#' @field or_criteria A list of Criteria objects for OR conditions.
#' @field not_criteria A list of Criteria objects for NOT conditions.
#'
#' @name Query
#' @export
Query <- setRefClass(
  "Query",
  fields = list(
    and_criteria = "list",
    or_criteria = "list",
    not_criteria = "list"
  ),
  methods = list(
    initialize = function(and_criteria = list(), or_criteria = list(), not_criteria = list()) {
      .self$and_criteria <- and_criteria
      .self$or_criteria <- or_criteria
      .self$not_criteria <- not_criteria
    },
    to_list = function() {
      return(list(
        and = lapply(.self$and_criteria, function(criteria) criteria$to_list()),
        or = lapply(.self$or_criteria, function(criteria) criteria$to_list()),
        not = lapply(.self$not_criteria, function(criteria) criteria$to_list())
      ))
    }
  )
)
